drivers/sensor/fcx_mldx5/fcx_mldx5.c

```diff
@@ -0,0 +1,495 @@
+static const char *const fcx_mldx5_cmds[] = {
+	[FCX_MLDX5_CMD_READ_STATUS] = "01",
+	[FCX_MLDX5_CMD_READ_O2_VALUE] = "02",
+	[FCX_MLDX5_CMD_SWITCH_SENSOR_ON_OFF] = "04",
+	[FCX_MLDX5_CMD_RESET] = "11",
+	[FCX_MLDX5_CMD_ERROR] = "EE",
+};
+
+static const uint8_t fcx_mldx5_cmds_data_len[] = {
+	[FCX_MLDX5_CMD_READ_STATUS] = 2,
+	[FCX_MLDX5_CMD_READ_O2_VALUE] = 5,
+	[FCX_MLDX5_CMD_SWITCH_SENSOR_ON_OFF] = 1,
+	[FCX_MLDX5_CMD_RESET] = 0,
+	[FCX_MLDX5_CMD_ERROR] = 2,
+};
+
+static const char *const fcx_mldx5_errors[] = {
+	[FCX_MLDX5_ERROR_CHECKSUM] = "checksum",
+	[FCX_MLDX5_ERROR_UNKNOWN_COMMAND] = "command",
+	[FCX_MLDX5_ERROR_PARAMETER] = "parameter",
+	[FCX_MLDX5_ERROR_EEPROM] = "eeprom",
+};
+
+static void fcx_mldx5_uart_flush(const struct device *uart_dev)
+{
+	uint8_t tmp;
+
+	while (uart_fifo_read(uart_dev, &tmp, 1) > 0) {
+		continue;
+	}
+}
+
+static uint8_t fcx_mldx5_calculate_checksum(const uint8_t *buf, size_t len)
+{
+	uint8_t checksum;
+	size_t i;
+
+	if (buf == NULL || len == 0) {
+		return 0;
+	}
+
+	checksum = buf[0];
+	for (i = 1; i < len; ++i) {
+		checksum ^= buf[i];
+	}
+
+	return checksum;
+}
+
+static int fcx_mldx5_frame_check_error(const struct fcx_mldx5_data *data, const char *command_sent)
+{
+	const uint8_t len = FCX_MLDX5_HEADER_LEN + fcx_mldx5_cmds_data_len[FCX_MLDX5_CMD_ERROR];
+	const char *command_error = fcx_mldx5_cmds[FCX_MLDX5_CMD_ERROR];
+	const char *command_received = &data->frame[FCX_MLDX5_CMD_INDEX];
+	const char *data_received = &data->frame[FCX_MLDX5_DATA_INDEX];
+	uint8_t error;
+
+	if (data->frame_len != len ||
+	    strncmp(command_error, command_received, FCX_MLDX5_CMD_LEN) != 0) {
+		return 0;
+	}
+
+	if (data_received[0] != 'E' || char2hex(data_received[1], &error) != 0 ||
+	    error >= ARRAY_SIZE(fcx_mldx5_errors)) {
+		LOG_WRN("Could not parse error value %.*s",
+			fcx_mldx5_cmds_data_len[FCX_MLDX5_CMD_ERROR], data_received);
+	} else {
+		LOG_WRN("Command '%s' received error '%s'", command_sent, fcx_mldx5_errors[error]);
+	}
+
+	return -EIO;
+}
+
+static int fcx_mldx5_frame_verify(const struct fcx_mldx5_data *data, enum fcx_mldx5_cmd cmd)
+{
+	const uint8_t frame_len = FCX_MLDX5_HEADER_LEN + fcx_mldx5_cmds_data_len[cmd];
+	const char *command = fcx_mldx5_cmds[cmd];
+	const char *command_received = &data->frame[FCX_MLDX5_CMD_INDEX];
+	uint8_t checksum;
+	uint8_t checksum_received;
+
+	if (fcx_mldx5_frame_check_error(data, command) != 0) {
+		return -EIO;
+	} else if (data->frame_len != frame_len) {
+		LOG_WRN("Expected command %s frame length %u not %u", command, frame_len,
+			data->frame_len);
+		return -EIO;
+	} else if (data->frame[FCX_MLDX5_STX_INDEX] != FCX_MLDX5_STX) {
+		LOG_WRN("No STX");
+		return -EIO;
+	} else if (strncmp(command, command_received, FCX_MLDX5_CMD_LEN) != 0) {
+		LOG_WRN("Expected command %s not %.*s", command, FCX_MLDX5_CMD_LEN,
+			command_received);
+		return -EIO;
+	} else if (data->frame[FCX_MLDX5_ETX_INDEX(data->frame_len)] != FCX_MLDX5_ETX) {
+		LOG_WRN("No ETX");
```

Comment:
Is there a reason not to use `LOG_ERR` for all of these error cases?

Reply:
No really. Will fix!

Reply:
> Is there a reason not to use `LOG_ERR` for all of these error cases?

Fixed. Also pipeline passes now.

Reply:
Ping @MaureenHelm :)